src/core/hle/kernel/thread.cpp

```diff
@@ -3,433 +3,972 @@
+    }
+
+    thread->waiting_on = objects;
+    for (auto obj : objects) {
+        obj->AddWaitingThread(Kernel::SharedFrom(thread));
+    }
+    thread->status = Thread::Status::WaitSyncAny;
+    Reschedule();
+    return thread->SyncOutput(RESULT_SUCCESS);
 }
 
-void Thread::ResumeFromWait() {
-    ASSERT_MSG(wait_objects.empty(), "Thread is waking up while waiting for objects");
+ResultCode ThreadManager::WaitIPC(std::vector<ObjectPtr> objects, s32* index) {
+    ASSERT(thread->status == Thread::Status::Running);
+    thread->sync_output = index;
 
-    switch (status) {
-    case ThreadStatus::WaitSynchAll:
-    case ThreadStatus::WaitSynchAny:
-    case ThreadStatus::WaitHleEvent:
-    case ThreadStatus::WaitArb:
-    case ThreadStatus::WaitSleep:
-    case ThreadStatus::WaitIPC:
-    case ThreadStatus::Dormant:
-        break;
+    // Check if we can sync immediately
+    auto active_obj = std::find_if(objects.begin(), objects.end(),
+                                   [this](const auto& obj) { return !obj->ShouldWait(thread); });
+    if (active_obj != objects.end()) {
+        s32 idx = static_cast<s32>(active_obj - objects.begin());
+        (*active_obj)->Acquire(thread);
 
-    case ThreadStatus::Ready:
-        // The thread's wakeup callback must have already been cleared when the thread was first
-        // awoken.
-        ASSERT(wakeup_callback == nullptr);
-        // If the thread is waiting on multiple wait objects, it might be awoken more than once
-        // before actually resuming. We can ignore subsequent wakeups if the thread status has
-        // already been set to ThreadStatus::Ready.
-        return;
-
-    case ThreadStatus::Running:
-        DEBUG_ASSERT_MSG(false, "Thread with object id {} has already resumed.", GetObjectId());
-        return;
-    case ThreadStatus::Dead:
-        // This should never happen, as threads must complete before being stopped.
-        DEBUG_ASSERT_MSG(false, "Thread with object id {} cannot be resumed because it's DEAD.",
-                         GetObjectId());
-        return;
-    }
-
-    wakeup_callback = nullptr;
-
-    thread_manager.ready_queue.push_back(current_priority, this);
-    status = ThreadStatus::Ready;
-    thread_manager.kernel.PrepareReschedule();
-}
+        if ((*active_obj)->GetHandleType() != HandleType::ServerSession) {
+            return thread->SyncOutput(RESULT_SUCCESS, idx);
+        }
+        auto server_session = dynamic_cast<ServerSession*>(active_obj->get());
+        return thread->SyncOutput(server_session->ReceiveIPCRequest(thread), idx);
+    }
 
-void ThreadManager::DebugThreadQueue() {
-    Thread* thread = GetCurrentThread();
-    if (!thread) {
-        LOG_DEBUG(Kernel, "Current: NO CURRENT THREAD");
-    } else {
-        LOG_DEBUG(Kernel, "0x{:02X} {} (current)", thread->current_priority,
-                  GetCurrentThread()->GetObjectId());
+    thread->waiting_on = objects;
+    for (auto obj : objects) {
```

Comment:
```suggestion
    for (auto& obj : objects) {
```

otherwise this can cause atomic reference count increment/decrement churn.